aws_lambda_powertools/utilities/idempotency/idempotency.py

```diff
@@ -0,0 +1,109 @@
+"""
+Primary interface for idempotent Lambda functions utility
+"""
+import logging
+from typing import Any, Callable, Dict
+
+from aws_lambda_powertools.middleware_factory import lambda_handler_decorator
+from aws_lambda_powertools.utilities.idempotency.persistence.base import STATUS_CONSTANTS, BasePersistenceLayer
+
+from ..typing import LambdaContext
+from .exceptions import AlreadyInProgressError, ItemAlreadyExistsError, ItemNotFoundError
+
+logger = logging.getLogger(__name__)
+
+
+def default_error_callback():
+    raise
+
+
+@lambda_handler_decorator
+def idempotent(
+    handler: Callable[[Any, LambdaContext], Any],
+    event: Dict[str, Any],
+    context: LambdaContext,
+    persistence_store: BasePersistenceLayer,
+) -> Any:
+    """
+    Middleware to handle idempotency
+
+    Parameters
+    ----------
+    handler: Callable
+        Lambda's handler
+    event: Dict
+        Lambda's Event
+    context: Dict
+        Lambda's Context
+    persistence_store: BasePersistenceLayer
+        Instance of BasePersistenceLayer to store data
+
+    Examples
+    --------
+    **Processes Lambda's event in an idempotent manner**
+        >>> from aws_lambda_powertools.utilities.idempotency import idempotent, DynamoDBPersistenceLayer
+        >>>
+        >>> persistence_store = DynamoDBPersistenceLayer(event_key="body", table_name="idempotency_store")
+        >>>
+        >>> @idempotent(persistence_store=persistence_store)
+        >>> def handler(event, context):
+        >>>     return {"StatusCode": 200}
+    """
+
+    try:
```

Comment:
It looks like only ItemAlreadyExistsError is being expected as possible error. As the persistence error can have other multiple exceptions, should these be managed too?

Reply:
For exceptions that aren't handled as ItemAlreadyExistsError, I think allowing the exception to bubble up is the correct approach in order to maintain the idempotency guarantees. We could wrap those exceptions into a custom exception and re-raise, but I don't think this adds much utility.

Reply:
Latest commit addresses this